uber/automated_emails.py

```diff
@@ -1,75 +1,290 @@
+            return False
+
+        for date_filter in self.when:
+            if not date_filter():
+                return False
+
+        return True
 
     def __repr__(self):
         return '<{}: {!r}>'.format(self.__class__.__name__, self.subject)
 
-    def prev(self, x, all_sent=None):
-        if all_sent:
-            return (x.__class__.__name__, x.id, self.ident) in all_sent
-        else:
-            with Session() as session:
-                return session.query(Email).filter_by(model=x.__class__.__name__, fk_id=x.id, ident=self.ident).first()
+    def _already_sent(self, model_inst):
+        """
+        Returns true if we have a record of previously sending this email to this model
+
+        NOTE: c.PREVIOUSLY_SENT_EMAILS is a cached property and will only update at the start of each daemon run.
+        """
+        return (model_inst.__class__.__name__, model_inst.id, self.ident) in c.PREVIOUSLY_SENT_EMAILS
 
-    def should_send(self, x, all_sent=None):
+    def send_if_possible(self, model_inst, raise_errors=False):
+        """
+        If it's OK to send an email of our category to this model instance (i.e. a particular Attendee) then send it.
+        """
         try:
-            return x.email and not self.prev(x, all_sent) and self.filter(x)
+            if self._should_send(model_inst):
+                self.really_send(model_inst)
         except:
-            log.error('unexpected error', exc_info=True)
+            log.error('error sending {!r} email to {}', self.subject, model_inst.email, exc_info=True)
+            if raise_errors:
+                raise
+
+    def _should_send(self, model_inst):
+        """
+        If True, we should generate an actual email created from our email category
+        and send it to a particular model instance.
+
+        This is determined based on a few things like:
+        1) whether we have sent this exact email out yet or not (previously_sent_emails)
+        2) whether the email category has been approved (approved_idents)
+        3) whether the model instance passed in is the same type as what we want to process
+        4) do any date-based filters exist on this email category? (i.e. send 7 days before magfest)
+        5) do any other filters exist on this email category? (i.e. only if attendee.staffing == true)
+
+        Example #1 of a model instance to check:
+          self.ident: "You {attendee.name} have registered for our event!"
+          model_inst:  class Attendee: id #4532, name: "John smith"
+
+        Example #2 of a model instance to check:
+          self.ident: "Your group {group.name} owes money"
+          model_inst:  class Group: id #1251, name: "The Fighting Mongooses"
+
+        :param model_inst: The model we've been requested to use (i.e. Attendee, Group, etc)
+
+        :return: True if we should send this email to this model instance, False if not.
+        """
+
+        if not isinstance(model_inst, self.model):
```

Comment:
I have both a dumb reason and a stylistic reason to do it this way.

The dumb reason is that it makes settings breakpoints way easier in the debugger when testing specific unit tests to make sure they are failing at the right points and getting past the earlier parts of this test.  This is purely for convenience and not really a great reason.

The stylistic reason is that I wanted this function to read like a checklist for the things that can prevent this from happening, where each line was one complete thought unrelated to the others.

We could mash it to together for a shorter version, but, I kinda like it this way if it's not a strong objection.

Alternatively we could later on down the line do something fancier like have a bunch of like ```@should_send``` properties that get called together (like what we do with ```@validation.Attendee```), which might also make it easier to unit test, add new properties, and read a little more compact.